simpl/dicts.py

```diff
@@ -0,0 +1,133 @@
+
+    :param target: a dict to write data to
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :param value: the value to write to the key
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    parts = path.split(separator)
+    current = target
+    for part in parts[:-1]:
+        if part not in current:
+            current[part] = current = {}
+        else:
+            current = current[part]
+    current[parts[-1]] = value
+
+
+def read_path(source, path, separator='/'):
+    """Read a value from a dict supporting a deep path as a key.
+
+    :param source: a dict to read data from
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    parts = path.strip(separator).split(separator)
+    current = source
+    for part in parts[:-1]:
+        if part not in current:
+            return
+        current = current[part]
+        if not isinstance(current, dict):
+            return
+    return current.get(parts[-1])
+
+
+def path_exists(source, path, separator='/'):
+    """Check a dict for the existence of a value given a path to it.
+
+    :param source: a dict to read data from
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    if path == separator and isinstance(source, dict):
+        return True
+    parts = path.strip(separator).split(separator)
+    if not parts:
+        return False
+    current = source
+    for part in parts:
+        if not isinstance(current, dict):
+            return False
+        if part not in current:
+            return False
+        current = current[part]
+    return True
+
+
+def merge_dictionary(dst, src, extend_lists=False):
+    """Recursively merge two dicts.
+
+    Hashes at the root level are NOT overwritten
+
+    Note: This updates dst.
+    """
+    stack = [(dst, src)]
+    while stack:
+        current_dst, current_src = stack.pop()
+        for key in current_src:
+            source = current_src[key]
+            if key not in current_dst:
+                current_dst[key] = source
+            else:
+                dest = current_dst[key]
+                if isinstance(source, dict) and isinstance(dest, dict):
+                    stack.append((dest, source))
+                elif isinstance(source, list) and isinstance(dest, list):
+                    merge_lists(dest, source, extend_lists=extend_lists)
+                else:
+                    current_dst[key] = source
+    return dst
+
+
+def merge_lists(dest, source, extend_lists=False):
+    """Recursively merge two lists.
+
+    This applies merge_dictionary if any of the entries are dicts.
```

Comment:
Can you explain the semantics of `extend_lists` in the docstring? Looking at the method signature, I had an idea about what this parameter means, but then when I looked at the code, my assumption was completely wrong. Looking at the test code in order to understand, I'm even more confused.

Some simple doctest examples would be ideal, to show the difference in behavior when `extend_lists` is `False` vs. when it is `True`.